File: frontend/src/features/expenses/services/expenses.js
import { apiFetch } from "@/lib/apiClient";

export const EXPENSE_DEPARTMENTS = [
  "Marketing",
  "Equipment",
  "Labor",
  "Studio / Venue",
  "Wardrobe / Styling",
  "Editing / Post",
  "Software",
  "Travel",
  "Food",
  "Distribution",
  "Merch",
  "Other",
];

export const normalizeExpense = (expense) => {
  if (!expense) {
    return null;
  }

  return {
    id: expense.id,
    userId: expense.userId,
    projectId: expense.projectId,
    name: expense.name,
    amount: Number(expense.amount ?? 0),
    department: expense.department,
    category: expense.category ?? null,
    description: expense.description ?? null,
    expenseDate: expense.expenseDate,
    vendor: expense.vendor ?? null,
    receiptUrl: expense.receiptUrl ?? null,
    createdAt: expense.createdAt,
    updatedAt: expense.updatedAt,
    project: expense.project ?? null,
  };
};

export const createExpense = async (payload, options = {}) => {
  if (!options.userId && !payload.userId) {
    throw new Error("You must be signed in to add an expense.");
  }

  if (!payload.projectId) {
    throw new Error("Project is required to add an expense.");
  }

  const body = await apiFetch(`/api/projects/${payload.projectId}/expenses`, {
    method: "POST",
    body: JSON.stringify({
      name: payload.name?.trim() || "",
      amount: payload.amount,
      department: payload.department,
      category: payload.category?.trim() || null,
      description: payload.description?.trim() || null,
      expenseDate: payload.expenseDate,
      vendor: payload.vendor?.trim() || null,
      receiptUrl: payload.receiptUrl?.trim() || null,
    }),
  });

  return {
    data: normalizeExpense(body.data),
    source: "api",
  };
};

export const getExpenses = async (options = {}) => {
  if (!options.userId) {
    throw new Error("You must be signed in to view expenses.");
  }

  if (!options.projectId) {
    return { data: [], source: "api" };
  }

  const params = new URLSearchParams();
  if (options.department && options.department !== "all") {
    params.set("department", options.department);
  }

  const qs = params.toString();
  const path = `/api/projects/${options.projectId}/expenses${qs ? `?${qs}` : ""}`;

  const body = await apiFetch(path);

  return {
    data: (body.data || []).map(normalizeExpense).filter(Boolean),
    source: "api",
  };
};

export const updateExpense = async (expenseId, payload, options = {}) => {
  const userId = options.userId ?? payload?.userId;

  if (!userId) {
    throw new Error("You must be signed in to update an expense.");
  }

  if (!expenseId) {
    throw new Error("Expense ID is required to update an expense.");
  }

  const projectId = payload.projectId || options.projectId;
  if (!projectId) {
    throw new Error("Project ID is required to update an expense.");
  }

  const body = await apiFetch(`/api/projects/${projectId}/expenses/${expenseId}`, {
    method: "PUT",
    body: JSON.stringify({
      name: payload.name,
      amount: payload.amount,
      department: payload.department,
      category: payload.category,
      description: payload.description,
      expenseDate: payload.expenseDate,
      vendor: payload.vendor,
      receiptUrl: payload.receiptUrl,
      projectId: payload.projectId,
    }),
  });

  return {
    data: normalizeExpense(body.data),
    source: "api",
  };
};

export const deleteExpense = async (expenseId, options = {}) => {
  if (!options.userId) {
    throw new Error("You must be signed in to delete an expense.");
  }

  if (!expenseId) {
    throw new Error("Expense ID is required to delete an expense.");
  }

  if (!options.projectId) {
    throw new Error("Project ID is required to delete an expense.");
  }

  await apiFetch(`/api/projects/${options.projectId}/expenses/${expenseId}`, {
    method: "DELETE",
  });

  return {
    data: { id: expenseId },
    source: "api",
  };
};

export const getBudgetSummary = async (projectId) => {
  const body = await apiFetch(`/api/projects/${projectId}/budget-summary`);
  return body.data;
};

File: backend/routes/expenses.py
from flask import Blueprint, jsonify, request, g
from flasgger import swag_from
from middleware.auth import require_auth
from services.expense_service import (
    create_expense,
    get_expenses,
    get_expense_by_id,
    update_expense,
    delete_expense,
    get_budget_summary,
)

expenses_bp = Blueprint("expenses", __name__)

REQUIRED_CREATE_FIELDS = ["name", "amount", "department", "expenseDate"]


@expenses_bp.route("/api/projects/<project_id>/expenses", methods=["POST"])
@require_auth
@swag_from("../docs/expenses_create.yml")
def create(project_id):
    body = request.get_json(silent=True) or {}

    missing = [f for f in REQUIRED_CREATE_FIELDS if not body.get(f)]
    if missing:
        return jsonify({"error": f"Missing required fields: {', '.join(missing)}"}), 400

    if not isinstance(body["amount"], (int, float)) or body["amount"] <= 0:
        return jsonify({"error": "amount must be a positive number"}), 400

    try:
        expense = create_expense(g.user_id, project_id, body)
        return jsonify({"data": expense}), 201
    except Exception as e:
        return jsonify({"error": str(e)}), 400


@expenses_bp.route("/api/projects/<project_id>/expenses", methods=["GET"])
@require_auth
@swag_from("../docs/expenses_list.yml")
def list_expenses(project_id):
    department = request.args.get("department")

    try:
        expenses = get_expenses(g.user_id, project_id, department=department)
        return jsonify({"data": expenses, "count": len(expenses)}), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 400


@expenses_bp.route("/api/projects/<project_id>/expenses/<expense_id>", methods=["GET"])
@require_auth
@swag_from("../docs/expenses_get.yml")
def get_one(project_id, expense_id):
    try:
        expense = get_expense_by_id(g.user_id, project_id, expense_id)
        if not expense:
            return jsonify({"error": "Expense not found"}), 404
        return jsonify({"data": expense}), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 404


@expenses_bp.route("/api/projects/<project_id>/expenses/<expense_id>", methods=["PUT"])
@require_auth
@swag_from("../docs/expenses_update.yml")
def update(project_id, expense_id):
    body = request.get_json(silent=True) or {}

    if "amount" in body:
        if not isinstance(body["amount"], (int, float)) or body["amount"] <= 0:
            return jsonify({"error": "amount must be a positive number"}), 400

    try:
        expense = update_expense(g.user_id, project_id, expense_id, body)
        if not expense:
            return jsonify({"error": "Expense not found"}), 404
        return jsonify({"data": expense}), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 400


@expenses_bp.route("/api/projects/<project_id>/expenses/<expense_id>", methods=["DELETE"])
@require_auth
@swag_from("../docs/expenses_delete.yml")
def delete(project_id, expense_id):
    try:
        deleted = delete_expense(g.user_id, expense_id)
        if not deleted:
            return jsonify({"error": "Expense not found"}), 404
        return jsonify({"message": "Expense deleted"}), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 400


@expenses_bp.route("/api/projects/<project_id>/budget-summary", methods=["GET"])
@require_auth
@swag_from("../docs/budget_summary.yml")
def budget_summary(project_id):
    try:
        summary = get_budget_summary(g.user_id, project_id)
        if not summary:
            return jsonify({"error": "Project not found"}), 404
        return jsonify({"data": summary}), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 400
